crates/lumina_shared/src/player/spaceship.rs

```diff
@@ -140,29 +142,125 @@ fn spaceship_movement(
+                    dash.direction = Vec2::ZERO;
+                    movement_stat.towards_linear_damping(35.0, damping_factor);
+                }
+            }
+
+            // Case 2: Dash is just starting, has energy, and not in cooldown
+            (false, true, true, true) => {
+                let current_direction = action
+                    .clamped_axis_pair(&PlayerAction::Move)
+                    .map(|axis| axis.xy())
+                    .unwrap_or_else(|| Vec2::new(0.0, 0.0))
+                    .normalize_or_zero();
+
+                if current_direction != Vec2::ZERO {
+                    dash.is_dashing = true;
+                    dash.duration = 0.7;
+                    dash.direction = current_direction;
+                    boost.energy -= dash.energy_cost;
+                    linear.0 = dash.direction * dash.speed;
+                    movement_stat.towards_linear_damping(0.0, damping_factor);
+                }
+            }
+
+            // Case 3: Dash is attempted but either in cooldown or insufficient energy
+            (false, true, has_energy, not_in_cooldown) => {
+                if !has_energy || !not_in_cooldown {
+                    if is_moving {
+                        // Calculate the current movement direction
+                        let current_direction = action
+                            .clamped_axis_pair(&PlayerAction::Move)
+                            .map(|axis| axis.xy())
+                            .unwrap_or_else(|| Vec2::ZERO)
+                            .normalize_or_zero();
+
+                        if current_direction != Vec2::ZERO {
+                            // Apply acceleration to maintain velocity
+                            movement_stat.towards_linear_acceleration(
+                                spaceship.linear_acceleration,
+                                acceleration_factor,
+                                deceleration_factor,
+                            );
+
+                            // Update velocity based on the current direction and acceleration
+                            linear.0 += current_direction
+                                * movement_stat.linear_acceleration
+                                * time.delta_seconds();
+                            linear.0 = linear.0.clamp_length_max(spaceship.max_linear_speed);
+                        }
+                    } else {
+                        // If not moving, apply braking behavior
+                        movement_stat.towards_linear_acceleration(
+                            0.0,
+                            acceleration_factor,
+                            deceleration_factor,
+                        );
+                    }
+
+                    // Apply damping to smoothly reduce velocity if necessary
+                    movement_stat.towards_linear_damping(spaceship.linear_damping, damping_factor);
+                }
+            }
```

Comment:
U might have noticed that the match case is abit too huge, some pointers would be to use state filtering.

In this example, u can first filter out if cooldown has reached, then filter out if there is enough energy.
A very simple example:
```rs
if cooldown > 0.0 {
    return;
}

if energy < required_energy {
    return;
}

// dash
```

Its best used this method in a separate system, but it is also possible to employ this in a separate function.

Reply:
As a rule of thumb when it comes to physics manipulation, do it at the end when u know the final velocity that u want, to prevent it being overwritten somewhere in the middle, this can be done using system ordering, u could create a middle representation of how values are being represented from one system to another.

also, the damping thing, I think it can be removed dy, since it was used in brakes xd